holoviews/core/dimension.py

```diff
@@ -87,6 +87,22 @@ def replace_dimensions(cls, dimensions, overrides):
         return replaced
 
 
+    def _filter_cache(self, dmap, kdims):
+        """
+        Pop keys from cache that don't match newly specified values
+        """
+        for kdim in kdims:
+            if kdim.values:
+                pos = kdims.index(kdim)
+                dropped = []
+                for key in dmap.data.keys():
+                    if key[pos] not in kdim.values:
+                        dropped.append(key)
+                for drop in dropped:
+                    del dmap.data[drop]
+        return dmap.data
```

Comment:
I'd invert these loops, instead of iterating over the keys once for each kdim this should be iterating over the keys once and filtering each key. Something like

```python
filtered = []
for key, value in dmap.data.items():
    skip = False
    for kd, v in zip(dmap.kdims, key):
          if kdim.values and v not in kdim.values:
               skip = True
     if not skip:
         filtered.append((k, v))
return filtered
```